### PR TITLE
feat(bots): implement proof of humanity bot

### PR DESCRIPTION
**Pull request by epiqueras**
# Description

This PR implements a new bot for the action callbacks of the [Proof Of Humanity smart contract](https://github.com/Proof-Of-Humanity/Proof-Of-Humanity/blob/master/contracts/ProofOfHumanity.sol).

This bot is a bit different from the others because it can leverage The Graph to more efficiently poll entities that match the contract's required conditions. I have structured the code into one module per action callback to clarify the relationship between the queries and the contract's required conditions.

E.g.

https://github.com/kleros/action-callback-bots/blob/68a27bc8a1f9400a68e155f999a728807966d6f0/src/bots/proof-of-humanity/change-state-to-pending.js#L3-L7

https://github.com/kleros/action-callback-bots/blob/68a27bc8a1f9400a68e155f999a728807966d6f0/src/bots/proof-of-humanity/change-state-to-pending.js#L33-L37

The PR also moves the repo to the new stable version of `web3`, because it caused compilation errors.

# How To Test This?

Run `npm start` and make sure the bot runs as expected in tandem with the others.

# Questions

I do not think these bots are resource-intensive enough to require us to run them on separate processes yet, but what is the throughput of the account sending the transactions?

Maybe it is time for each bot to get an individual account?

### CODE REVIEW

File: src/bots/proof-of-humanity/change-state-to-pending.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
+const { gql } = require('graphql-request')
+
+// changeStateToPending(address _submissionID, address[] calldata _vouches)
+// require(submission.status == Status.Vouching, "Wrong status");
+// require(round.hasPaid[uint(Party.Requester)], "Requester didn't pay his fees");
+// require(request.vouches.length >= requiredNumberOfVouches, "Not enough valid vouches");
```
Comment (satello) [inline src/bots/proof-of-humanity/change-state-to-pending.js:6]: Can these comments be removed?

File: src/bots/proof-of-humanity/change-state-to-pending.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            }
+          }
+        }
+      }
+    `
+  )
+
+  const submissionsWithVouches = await Promise.all(
+    submissions
+      // require(round.hasPaid[uint(Party.Requester)], "Requester didn't pay his fees");
```
Comment (satello) [inline src/bots/proof-of-humanity/change-state-to-pending.js:34]: Remove or make more readable

File: src/bots/proof-of-humanity/change-state-to-pending.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+          {
+            id: [submission.id]
+          }
+        )).submissions.map(submission => submission.id)
+      }))
+  )
+
+  return (
+    submissionsWithVouches
+      // require(request.vouches.length >= requiredNumberOfVouches, "Not enough valid vouches");
```
Comment (satello) [inline src/bots/proof-of-humanity/change-state-to-pending.js:57]: Remove or make more readable

File: src/bots/proof-of-humanity/execute-request.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
+const { gql } = require('graphql-request')
+
+// executeRequest(address _submissionID)
+// require(now - request.lastStatusChange > challengePeriodDuration, "Can't execute yet");
+// require(!request.disputed, "The request is disputed");
+// require(submission.status == Status.PendingRegistration || submission.status == Status.PendingRemoval);
```
Comment (satello) [inline src/bots/proof-of-humanity/execute-request.js:6]: Remove?

File: src/bots/proof-of-humanity/execute-request.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+          }
+        }
+      }
+    `
+  )
+
+  return (
+    [...pendingRegistration, ...pendingRemoval]
+      // require(!request.disputed, "The request is disputed");
+      // require(now - request.lastStatusChange > challengePeriodDuration, "Can't execute yet");
```
Comment (satello) [inline src/bots/proof-of-humanity/execute-request.js:43]: Remove

File: src/bots/proof-of-humanity/process-vouches.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
+const { gql } = require('graphql-request')
+
+// processVouches(address _submissionID, uint _requestID, uint _iterations)
+// require(request.resolved, "Submission should be resolved");
+// request.penaltyIndex = endIndex.toUint32();
```
Comment (satello) [inline src/bots/proof-of-humanity/process-vouches.js:5]: Remove?

File: src/bots/proof-of-humanity/process-vouches.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+          }
+        }
+      }
+    `
+  )
+
+  return submissions.flatMap(({ id, requests }) =>
+    requests
+      // require(request.resolved, "Submission should be resolved");
+      // request.penaltyIndex = endIndex.toUint32();
```
Comment (satello) [inline src/bots/proof-of-humanity/process-vouches.js:28]: Remove

File: src/bots/proof-of-humanity/withdraw-fees-and-rewards.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
+const { gql } = require('graphql-request')
+
+// withdrawFeesAndRewards(address payable _beneficiary, address _submissionID, uint _requestID, uint _challengeID, uint _round)
+// require(request.resolved, "Submission should be resolved");
```
Comment (satello) [inline src/bots/proof-of-humanity/withdraw-fees-and-rewards.js:4]: Remove

File: src/bots/proof-of-humanity/withdraw-fees-and-rewards.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+          }
+          contributor
+        }
+      }
+    `
+  )
+
+  return (
+    contributions
+      // require(request.resolved, "Submission should be resolved");
```
Comment (satello) [inline src/bots/proof-of-humanity/withdraw-fees-and-rewards.js:39]: remove

**Review by satello — COMMENTED** [review]
Is there a reason you have commented out require statements throughout?

**Comment by satello** [comment]
We don't hit gas limits in our batched transactions besides when we are executing and redistributing disputes. These are currently not happening more than 1-2 a day so we are not running into issues with large queues of transactions with significant delays yet. I think we can hold off on splitting out accounts/running on different processes for now. We are having a hard enough time keeping one bot account funded at the moment. If Proof of Humanity were to see enough traffic to need a dedicated PK and batching process we can make the change at that time.

File: src/bots/proof-of-humanity/change-state-to-pending.js (hunk 1 above)
Comment (mtsalenc) [reply src/bots/proof-of-humanity/change-state-to-pending.js:6]: > Is there a reason you have commented out require statements throughout?

I think the idea was to document when this will run.

Comment (satello) [reply src/bots/proof-of-humanity/change-state-to-pending.js:6]: Let's put them in a more readable format.

Comment (epiqueras) [reply src/bots/proof-of-humanity/change-state-to-pending.js:6]: Yeah, I explained it in the description and gave an example:

>I have structured the code into one module per action callback to clarify the relationship between the queries and the contract's required conditions.

What format do you suggest?

The idea is that this makes it super easy to copy and search in the contract where the logic is.

**Comment by epiqueras** [comment]
>We don't hit gas limits in our batched transactions besides when we are executing and redistributing disputes. These are currently not happening more than 1-2 a day so we are not running into issues with large queues of transactions with significant delays yet. I think we can hold off on splitting out accounts/running on different processes for now. We are having a hard enough time keeping one bot account funded at the moment. If Proof of Humanity were to see enough traffic to need a dedicated PK and batching process we can make the change at that time.

Sounds good. When that time comes, we should look at making a funding contract.

File: src/bots/proof-of-humanity/change-state-to-pending.js (hunk 1 above)
Comment (satello) [reply src/bots/proof-of-humanity/change-state-to-pending.js:6]: I think the part that makes it hard to read is the error messages. As the error message are the negation of the statement when I read a line I have to do a little mental gymnastics each time to figure out exactly what it is saying. And overall reading through code is more difficult than a `docstring` should be. It would be much easier to follow if the conditions are in plain english. You can leave the function signature at the top if you want it to be easily searchable in the smart contract.

e.g.
`
changeStateToPending(address _submissionID, address[] calldata _vouches)
Change the state of a submission to pending. Submissions must have a status of Vouching, have paid the fees, and have at least the required number of vouches. 
`

**Review by satello — APPROVED** [review]
